### PR TITLE
Use paragonie/random-compat

### PR DESCRIPTION
**Pull request by cweagans**
`paragonie/random-compat` is a polyfill for `random_bytes()`. Leveraging it allows phpseclib to dump a lot of code and not lose any functionality.

Note that this PR also includes https://github.com/phpseclib/phpseclib/pull/844

### CODE REVIEW

File: phpseclib/Crypt/Random.php [hunk 1]
```diff
@@ -54,66 +54,21 @@ class Random
...
-                // as normal. Note, however, that if we don't have a sufficient source of randomness for
-                // random_bytes(), most of the other calls here will fail too, so we'll end up using
-                // the PHP implementation.
-            }
+        // paragonie/random_compat defines random_bytes() for us if we're using
+        // PHP < 7.0.0. If it can't find any good sources of randomness,
+        // random_bytes() is defined as a function that just throws an exception.
+        try {
+            return \random_bytes($length);
+        } catch (\Throwable $e) {
```
Comment (GrahamCampbell) [inline phpseclib/Crypt/Random.php]: does throwable exist on php 5.6 though?

**Comment by terrafrost** [comment]
https://raw.githubusercontent.com/paragonie/random_compat/master/lib/random_bytes_openssl.php

Per the comments of https://github.com/phpseclib/phpseclib/blob/5745c8a5beed2d98625f0bbc3a238a8a3387c49d/phpseclib/Crypt/Random.php#L75 there's "possible blocking behavior" in openssl_random_pseudo_bytes prior to PHP 5.3.4 that random_bytes_openssl doesn't take into account. Although the comments in phpseclib's Crypt/Random.php don't elaborate I believe the issue was due OpenSSL's `RAND_pseudo_bytes` trying to gather entropy from the screen and blocking until sufficient entropy was gathered. 

We can download PHP 5.3.2 for Windows here:

http://windows.php.net/downloads/qa/test/5.3.2-final/

File: phpseclib/Crypt/Random.php (hunk 1 above)
Comment (cweagans) [inline phpseclib/Crypt/Random.php]: random-compat has an error polyfill: https://github.com/paragonie/random_compat/blob/master/lib/error_polyfill.php

**Comment by cweagans** [comment]
openssl is prioritized last in random-compat for a reason - see last paragraph of https://github.com/paragonie/random_compat/blob/master/ERRATA.md.

If somebody is running a PHP app on a Windows server with a (very) old version of PHP installed and nothing else other than openssl, it will *still* use random_bytes() backed by the `com_dotnet` extension first, which is enabled by default on PHP > 5.3.15.

File: phpseclib/Crypt/Random.php (hunk 1 above)
Comment (GrahamCampbell) [inline phpseclib/Crypt/Random.php]: Nice :)

Comment (GrahamCampbell) [inline phpseclib/Crypt/Random.php]: Hmm, well, assuming we're 100% certain all exceptions will defientely actually implement throwable on 5.6.

Comment (cweagans) [inline phpseclib/Crypt/Random.php]: The exceptions that the random-compat lib throws definitely implement Throwable (also in the error polyfill). In 7.0, *all* exceptions implement throwable.

Comment (GrahamCampbell) [inline phpseclib/Crypt/Random.php]: Yeh. I was just checking to be sure there's zero chance of an exception coming out of there that might not implement throwable.

Comment (cweagans) [inline phpseclib/Crypt/Random.php]: Fair enough. There's definitely a 0% chance.

**Comment by cweagans** [comment]
Note also that if the PHP 5.4 requirement lands, the openssl version check becomes even less of an issue.

File: phpseclib/Crypt/Random.php (hunk 1 above)
Comment (cweagans) [inline phpseclib/Crypt/Random.php]: So. Turns out I was wrong. https://github.com/paragonie/random_compat/pull/47

We'd need to catch both `\Throwable` and `\Exception` in this block.

Comment (GrahamCampbell) [inline phpseclib/Crypt/Random.php]: lol

Comment (cs278) [inline phpseclib/Crypt/Random.php]: See https://github.com/paragonie/random_compat/pull/56 you are better off catching `\Error` and `\Exception` that way you get the same execution path on PHP 5 and PHP 7.

**Comment by bantu** [comment]
@terrafrost What is the status here?

File: composer.json [hunk 2]
```diff
@@ -51,7 +51,8 @@
         }
     ],
     "require": {
-        "php": ">=5.3.3"
+        "php": ">=5.3.3",
+        "paragonie/random_compat": "~1.0"
```
Comment (GrahamCampbell) [inline composer.json]: maybe `~1.1` would be a better idea here?

**Comment by terrafrost** [comment]
> @terrafrost What is the status here?

I'm thinking about just letting this sit until a decision has been made on adding Composer as a pre-requisite, which kinda ties into #763. And since master is a good ways off, anyway, I'm in no particular rush to make a decision on this atm..

**Comment by terrafrost** [comment]
This has been merged with #958

**Pull request closed without merging**